core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/recipe/RecipeExecutionFailureCollector.java

```diff
@@ -1,69 +1,44 @@
 public class RecipeExecutionFailureCollector {
 
+    private static final Pattern PHASE_PATTERN = Pattern.compile(".*/opt/scripts/recipe-runner\\.sh ([a-zA-Z0-9-]+) (([a-zA-Z0-9-]+)).*", Pattern.DOTALL);
+
     public boolean canProcessExecutionFailure(CloudbreakOrchestratorException exception) {
         return !getNodesWithErrors(exception).isEmpty() && exception.getMessage().contains("/opt/scripts/recipe-runner.sh ");
     }
 
-    public Set<RecipeExecutionFailure> collectErrors(CloudbreakOrchestratorException exception, Map<HostGroup, List<RecipeModel>> hostgroupToRecipeMap,
-            Set<InstanceGroup> instanceGroups) {
-
+    public List<RecipeFailure> collectErrors(CloudbreakOrchestratorException exception) {
         Multimap<String, String> nodesWithErrors = getNodesWithErrors(exception);
 
         if (nodesWithErrors.isEmpty()) {
             throw new CloudbreakServiceException("Failed to collect recipe execution failures. Cause exception contains no information.", exception);
         }
 
-        return nodesWithErrors.asMap().entrySet().stream().flatMap((Entry<String, Collection<String>> nodeWithErrors) -> {
-
-            Map<String, Optional<String>> errorsWithPhase = nodeWithErrors.getValue().stream()
-                    .collect(Collectors.toMap(error -> error, this::getInstallPhase, (phase1, phase2) -> phase1.or(() -> phase2)));
-
-            return errorsWithPhase.entrySet().stream()
-                    .filter(errorWithPhase -> errorWithPhase.getValue().isPresent())
-                    .map(this::errorWithPhaseToRecipeName)
-                    .collect(Collectors.toMap(
-                            recipeName -> recipeName,
-                            recipeName -> getPossibleFailingHostgroupsByRecipeName(hostgroupToRecipeMap, recipeName),
-                            (hg1, hg2) -> Stream.concat(hg1.stream(), hg2.stream()).collect(Collectors.toSet()))
-                    )
-                    .entrySet().stream()
-                    .flatMap((Entry<String, Set<String>> recipeWithHostgroups) -> recipeWithHostgroups.getValue().stream()
-                            .map((String hostGroup) -> {
-                                Optional<Recipe> recipe = getRecipeOfHostgroupByRecipeName(hostgroupToRecipeMap, recipeWithHostgroups.getKey(), hostGroup);
-                                Optional<InstanceMetaData> instanceMetaData = getInstanceMetaDataByHost(instanceGroups, nodeWithErrors.getKey(), hostGroup);
-                                if (recipe.isPresent() && instanceMetaData.isPresent()) {
-                                    return new RecipeExecutionFailure(recipe.get(), instanceMetaData.get());
-                                }
-                                return null;
-                            })
-                            .filter(Objects::nonNull)
-                    );
-        }).collect(Collectors.toSet());
+        List<RecipeFailure> failures = nodesWithErrors.asMap().entrySet().stream()
```

Comment:
Nice and clean implementation. Was there a salt change before, that made this possible?

Reply:
Thanks. 
Not really, I just got the idea to parse recipe names from the Error message aswel with regex, and removed some optionals.

Reply:
cool